notes/structural-triples.md

```diff
@@ -0,0 +1,310 @@
+```
+
+This does not solved a schema migration problem, but a best practice at least provides means to iterate over data model. In a decentralized settings however two conflicting meanings can be assigned a names like `album` and there is no clear path to reconsiling them.
+
+### Semantic datalog
+
+[Datomic] and [RDFOX] break away from the norm and embrace semantic triples for modeling data. Here is an example of the same ancestor rule in datomic notation
+
+```clj
+[(ancestor ?person ?ancestor)
+ [?person :person/parent ?ancestor]]
+[(ancestor ?person ?ancestor)
+ [?person :person/parent ?parent]
+ (ancestor ?parent ?ancestor)]
+```
+
+And here is the equivalent in RDFOX syntax which is almost identical to one used by datomic.
+
+```sparql
+[?person, :ancestor, ?ancestor] :-
+  [?person, :parent, ?ancestor] .
+[?person, :ancestor, ?ancestor] :-
+  [?person, :parent, ?parent],
+  [?parent, :ancestor, ?acestor] .
+```
+
+#### Embedded wisdom
+
+It may not be immediately obvious but they do something very interesting, they lifted conventional wisdom from a best practice to an only possibly practice.
+
+- Facts are modeled as semantic triples in form of `[entity, attribute, value]` (a.k.a `[subject, predicate, object]`) ingraining best pracite about `ID`s.
+
+- Facts as semantic triples take composibility to an extreme, because all records end up breaking down to smallest possible units from which everything else is composed.
+
+Previously we end up defining `song_in_album` facts to capture relation as a semantic triple `[?song :song/album ?album]` which enabled us to express extension for `Track` data model and in turn allowed us to define a `track_v2` as composition of two. In type notation it could corresponds to rougly following
+
+```ts
+type Track = {
+  artist: Artist
+  title: string
+  duration: number
+}
+
+type Artist = {
+  name: string
+}
+
+type Album = {
+  name: string
+}
+
+type TrackAlbum = {
+  album: Album
+}
+
+type TrackV2 = Track & TrackAlbum
+```
+
+🤔 However it is good to ponder how things would have turned out if we have started track that accounted for `album`  from the begining, most likely all tracks would have had `artist`, `title`, `album`, and `duration`, but than how would we have stored information about songs that we did not know an `album` or a `duration` for ?
+
+#### Subetyping
+
+In type systems you could define a supertype containing subset of the subtype in fact that is what we end up defining in our `song` rule in datalog example. In other words we broke down model into smaller parts and then composed `track` from them.
+
+That is effectively what triples do they just take it to a logical extreme. This effectively reduces need for schema migration because our tables end up just maps of `entity -> value` where name of the table serves as an `attribute`.
+
+Everything else is just composition (or a query over those tables), therefor instead of migartion we can simply define a new query.
```

Comment:
```suggestion
Everything else is just composition (or a query over those tables), therefor instead of migration we can simply define a new query.
```